contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,348 @@
+ 
+ However, by some extension to this effort, we would like to improve scalability
+of the system, by significantly reducing amount of api calls coming from kubelets.
+As of now, to avoid situation that kubelet is watching all secrets/configmaps/...
+in the system, it is not using watch. Instead of that, it is retrieving indidual
+objects, by sending individual GET requests. However, it is sending those requests
+periodically to enable automatic updates of mounted secrets/configmaps/...
+In large clusters, this is generating huge unnecessary load, as this load in
+principle should be watch-based. We would like to address this together with
+solving the authorization issue.
+
+
+# Proposal
+
+ In this proposal, we're not focusing on how exactly security should be done.
+We're just sketching very high level approach and exact authorization mechanism
+should be discussed separately.
+
+ At the high level, what we would like to achieve is to enable LIST and WATCH
+requests to support more sophisticated filtering (e.g. we would like to be able
+to somehow ask for all Secrets attached to all pods bound to a given node in a
+single LIST or WATCH request). We obviously want to be able to authorize all
+other types of requests (in particular GETs), so that the design has to be
+consistent in all those cases.
+
+ To solve this problem, we propose to introduce the idea of `bulk watch ` and
+`bulk get ` (and in the future other bulk operations). This idea was already
+appearing in the community and now it seems to be a good time to proceed with it.
+
+ To allow Kubelet to use watch, we also need to periodically verify ACLs of all
+watches. Whenever a user loses access to a given resource, the watch should be
+close within some bounded time. The rationale behind this requirement is that
+by using (bulk) watch we can't regress the current behavior where a polling
+Kubelet would eventually be unable to access a secret.
+
+TODO: I think this isn't specific to bulk watch and we need to solve this
+problem also in regular watch. For example watcher watching just a single
+secret may also lose access to it and this watch should be closed in this case.
+
+ We would like the API to be generic enough so that it will be useful for many
+different usecases, not just for this particular one. As a result, below we are
+describing requirements that the solution has to meet to satisfy our needs:
+- a single bulk request has to support multiple resources (e.g. get a node and
+all pods associated with it)
+- if the bulk request filters asks only for object of a single resource type,
+the returned result should ideally be comparible with what is returned from
+the regular API
+- the API has to support aggregation if different resource types are server by
+different apiservers
+- ACLs of watches have to be periodically verified and automatically closed if
+ACLs no longer allow for this watch.
+- at some point in the future we want clients to be able to alter their watch
```

Comment:
I guess you could argue that dynamic membership change will happen anyway, so in a controller we want to have an `UpdateSubscriptions([]subscriptions)` call so that you can resync the set of watches that are open.  I.e. node deletion, pod deletion, new scheduled pod will all change the set of open watches.

The interesting thing here is the tension between what a client wants (to build a single list and update the watches in bulk) and what the machinery of that client would want (to deliver incremental pub/sub calls to the server).

I kind of agree that the full re-watch behavior seems really unusual.  As in, it seems very inefficient and something that anyone undergoing a reasonable amount of watch change would want to avoid.